src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/services/AuditS3ClientTest.kt

```diff
@@ -36,27 +40,85 @@ class AuditS3ClientTest {
     val putObjectRequest = putObjectRequestCaptor.value
     assertThat(putObjectRequest.bucket()).isEqualTo("bucketName")
-    assertThat(putObjectRequest.key()).isEqualTo("year=2020/month=12/day=31/user=testUser/07c867b1fb88d84fdcb16090619da3da.json")
-    assertThat(String(requestBodyCaptor.value.contentStreamProvider().newStream().readAllBytes(), StandardCharsets.UTF_8)).isEqualTo(auditEventJsonString)
+    assertThat(putObjectRequest.key()).isEqualTo("year=2020/month=12/day=31/user=testUser/$auditEventId.parquet")
+    verifyUploadedFileContents(auditEvent, putObjectRequest)
+  }
+
+  @Test
+  fun `save should throw exception when ID is missing`() {
+    val auditEvent = HMPPSAuditListener.AuditEvent(
+      what = "some event",
+      `when` = LocalDateTime.of(2020, 12, 31, 13, 30).toInstant(ZoneOffset.UTC),
+      operationId = "some operation ID",
+      subjectId = "some subject ID",
+      subjectType = "some subject type",
+      correlationId = "some correlation ID",
+      who = "testUser",
+      service = "some service",
+      details = "some details",
+    )
+
+    val exception = assertThrows<IllegalArgumentException> { auditS3Client.save(auditEvent) }
+    assertThat(exception.message).isEqualTo("ID cannot be null")
+    then(s3Client).shouldHaveNoInteractions()
+  }
+
+  private fun verifyUploadedFileContents(
+    auditEvent: HMPPSAuditListener.AuditEvent,
+    putObjectRequest: PutObjectRequest,
+  ) {
+    val requestBody = requestBodyCaptor.value
+    val parquetBytes = requestBody.contentStreamProvider().newStream().readAllBytes()
+
+    val tempFile = Files.createTempFile("test-parquet", ".parquet").toFile()
+    Files.write(tempFile.toPath(), parquetBytes)
+
+    val reader = AvroParquetReader.builder<GenericRecord>(Path(tempFile.toURI())).build()
+
+    val record: GenericRecord = reader.read()
+
+    assertThat(record.get("id").toString()).isEqualTo(auditEvent.id.toString())
+    assertThat(record.get("what").toString()).isEqualTo(auditEvent.what)
+    assertThat(record.get("when").toString()).isEqualTo(auditEvent.`when`.toString())
+    assertThat(record.get("operationId").toString()).isEqualTo(auditEvent.operationId)
+    assertThat(record.get("subjectId").toString()).isEqualTo(auditEvent.subjectId)
+    assertThat(record.get("subjectType").toString()).isEqualTo(auditEvent.subjectType)
+    assertThat(record.get("correlationId").toString()).isEqualTo(auditEvent.correlationId)
+    assertThat(record.get("who").toString()).isEqualTo(auditEvent.who)
+    assertThat(record.get("service").toString()).isEqualTo(auditEvent.service)
+    assertThat(record.get("details").toString()).isEqualTo(auditEvent.details)
+
+    val expectedMd5 = MessageDigest.getInstance("MD5").digest(parquetBytes)
+    val expectedMd5Base64 = Base64.getEncoder().encodeToString(expectedMd5)
+    assertThat(putObjectRequest.contentMD5()).isEqualTo(expectedMd5Base64)
```

Comment:
Might be worth added another test case to verify the temp file doesn't exist after its marshalled the event